recipes/cern-root/all/conanfile.py

```diff
@@ -0,0 +1,330 @@
+import os
+import shutil
+import stat
+from glob import glob
+
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+
+
+class PythonOption:
+    OFF = "off"
+    SYSTEM = "system"
+    # in future we may allow the user to specify a version when
+    # libPython is available in Conan Center Index.
+    # FIXME: add option to use CCI Python package when it is available
+    ALL = [OFF, SYSTEM]
+
+
+required_conan_version = ">=1.29.1"
+
+
+class CernRootConan(ConanFile):
+    name = "cern-root"
+    # version format is intentional, ROOT does not follow strict SemVer.
+    # see: https://root.cern/about/versioning/
+    license = "LGPL-2.1-or-later"  # of ROOT itself, the recipe is under MIT license.
+    homepage = "https://root.cern/"
+    # ROOT itself is located at: https://github.com/root-project/root
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "CERN ROOT data analysis framework."
+    topics = ("data-analysis", "physics")
+    settings = ("os", "compiler", "build_type", "arch")
+    options = {
+        # Don't allow static build as it is not supported
+        # see: https://sft.its.cern.ch/jira/browse/ROOT-6446
+        # FIXME: shared option should be reinstated when hooks issue is resolved
+        # (see: https://github.com/conan-io/hooks/issues/252)
+        # "shared": [True],
+        "fPIC": [True, False],
+        "python": PythonOption.ALL,
+    }
+    default_options = {
+        # "shared": True,
+        "fPIC": True,
+        # default python=off as there is currently no libpython in Conan center
+        "python": PythonOption.OFF,
+    }
+    generators = ("cmake", "cmake_find_package")
+    requires = (
+        "opengl/system",
+        "libxml2/2.9.10",
+        "glu/system",
+        "xorg/system",
+        "sqlite3/3.34.0",
+        "libjpeg/9d",
+        "libpng/1.6.37",
+        "libcurl/7.74.0",
+        "pcre/8.44",
+        "xz_utils/5.2.5",
+        "zstd/1.4.8",
+        "lz4/1.9.3",
+        "glew/2.1.0",
+        "openssl/1.1.1i",
+        "fftw/3.3.8",
+        "cfitsio/3.490",
+        "tbb/2020.3",
+        "libpng/1.6.37",
+    )
+
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    @property
+    def _minimum_cpp_standard(self):
+        return 11
+
+    @property
+    def _minimum_compilers_version(self):
+        return {
+            "Visual Studio": "16.1",
+            "gcc": "4.8",
+            "clang": "3.4",
+            "apple-clang": "5.1",
+        }
+
+    def configure(self):
+        del self.settings.compiler.libcxx
```

Comment:
Are you sure root is c only?
I browsed through its github project and found some xxx files, that are included by default.
E.g. https://github.com/root-project/root/tree/master/core/textinput/src

Reply:
It's due to the fact the injected values from Conan break the fragile very explicit settings from upstream

Reply:
ROOT is C++.

I explained the reason for this commit here: <https://github.com/conan-io/conan-center-index/pull/3732#discussion_r567484902>.

The problem is that `conan_set_libcxx()` (called by `conan_basic_setup()`) causes the build to fail. Calling `del self.settings.compiler.libcxx` prevents this.

Reply:
Thanks for the comments.
What os/compiler/version/libcxx is causing problems?

Reply:
Both clang and gcc on Linux if libcxx is set to anything other than `libstdc++11`.

Reply:
Did we try allowing the setting and throwing invalid configuration when it's not `libstdc++11` ( too many iterations )

Reply:
No worries. It is a very long pull request. I am also losing track of things. :)

Yes, that will also work. That would be option (2) discussed in the comment <https://github.com/conan-io/conan-center-index/pull/3732#discussion_r567484902>.

I checked that that works locally and pushed it to this PR to check that it passes the CI checks (this was in c68ef3e). @SpaceIm pointed out in <https://github.com/conan-io/conan-center-index/pull/3732#discussion_r567450575> that this commit will fail if the compiler is Visual Studio.

I wasn't sure which of the 2 options was the best solution. 

If people think that raising `ConanInvalidConfiguration` is a better solution, I can revert 5dc549467 and reinstate c68ef3e and I can add something to handle other compilers.

Reply:
You can revert, but use this kind of condition instead `if libcxx and libcxx != "libstdc++11"`

Reply:
The link error (on my system, gcc9+libstdc++) is:
```[ 77%] Linking CXX executable ../bin/rootcling
/usr/bin/ld: ../lib/libCling.so.6.22.06: undefined reference to `clang::DeclarationName::getAsString[abi:cxx11]() const'
/usr/bin/ld: ../lib/libCling.so.6.22.06: undefined reference to `llvm::Twine::str[abi:cxx11]() const'
/usr/bin/ld: ../lib/libCling.so.6.22.06: undefined reference to `clang::getClangRevision[abi:cxx11]()'
```
Which means we're mixing the 2 abi's.
When I checked the exported symbols from `libCling.so`, found a lot of symbols with `cxx11` in its name.
I think the reason lies with some [cmake external projects](https://cmake.org/cmake/help/latest/module/ExternalProject.html).
e.g. for clad, my log contains (reduced a bit to make it readable):
```
[ 76%] Creating directories for 'clad'
[ 76%] Performing download step (git clone) for 'clad'
[ 76%] No patch step for 'clad'
[ 76%] No update step for 'clad'
[ 76%] Performing configure step for 'clad'
cd /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-build && /usr/bin/cmake -P /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-configure-Release.cmake
-- clad configure command succeeded.  See also /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-configure-*.log
cd /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-build && /usr/bin/cmake -E touch /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-configure
[ 76%] Performing build step for 'clad'
cd /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-build && /usr/bin/cmake -Dmake=/usr/bin/gmake -P /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-build-Release.cmake
-- clad build command succeeded.  See also /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-build-*.log
cd /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-build && /usr/bin/cmake -E touch /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-build
[ 76%] Performing install step for 'clad'
cd /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-build && /usr/bin/cmake -P /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-install-Release.cmake
-- clad install command succeeded.  See also /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-install-*.log
cd /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-build && /usr/bin/cmake -E touch /home/maarten/.conan/data/cern-root/v6-22-06/_/_/build/b8ad9631262c769e1d8d13dad4eddaa104e440bb/build_subfolder/interpreter/llvm/src/tools/cling/tools/plugins/clad/clad-prefix/src/clad-stamp/clad-install
[ 76%] Completed 'clad'
```

This shows two problem:
- it fetches external projects, which is not (version) controlled by conan.
- the `external projects` are not configured correctly to pass various cmake variables and/or properties.
In this case, the `COMPILE_DEFINITIONS` property is not propagated.

Reply:
e.g. this command: https://github.com/root-project/root/blob/730d24b73296a7d0b78d33588a5c9fdbb80e6731/interpreter/cling/tools/plugins/clad/CMakeLists.txt#L49-L76

Reply:
Adding `"clad": False` to the cmake definitions and removing `del self.settings.compiler.libcxx` made it build and link on my system for libstdc++.


The test_package now fails with the following error:
```
[100%] Linking CXX executable testrootdictionaries
/usr/bin/ld: libEvent.so: undefined reference to `TROOT::RegisterModule(char const*, char const**, char const**, char const*, char const*, void (*)(), std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&, char const**, bool)'
```

I didn't check whether there are some other hidden `ExternalProject_Add`'s lingering,
but there probably are.

Reply:
Ok, it appears the test_package was not using the `cmake` generator.
By adding `cmake` to the generators, including `conanbuildinfo.cmake` and calling `conan_basic_setup()` in the `CMakeLists.txt` and running `bin/testrootdirectories`, it fully works.

Reply:
@davehadley Can you please add a comment in the main `conanfile.py` for the reason why you're disabling `clad`?

Reply:
Great thank you!

Yes. I will add this.

Reply:
Thank you for your help. This is implemented in d0a1795 and fd02ab0.